Gradle test-retry-gradle-plugin doesn't work with Kotest.  E: The following test methods could not be retried, which is unexpected.

Project to reproduce [kotest-retry.zip](https://github.com/kotest/kotest/files/13872038/kotest-retry.zip)

Tested with:
Gradle 8.4
Kotest 5.8.0
Kotlin 1.9.22
JDK 11

Error:
```
Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':test'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:148)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:282)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:146)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:134)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:74)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:199)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:78)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:42)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:331)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:318)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.lambda$execute$0(DefaultTaskExecutionGraph.java:314)
        at org.gradle.internal.operations.CurrentBuildOperationRef.with(CurrentBuildOperationRef.java:80)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:314)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:303)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:463)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:380)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
Caused by: java.lang.IllegalStateException: The following test methods could not be retried, which is unexpected. Please file a bug report at https://github.com/gradle/test-retry-gradle-plugin/issues
   com.example.ApplicationKotestTest#should pass once per 5 times

        at org.gradle.testretry.internal.executer.RetryTestExecuter.failWithNonRetriedTestsIfAny(RetryTestExecuter.java:144)
        at org.gradle.testretry.internal.config.TestTaskConfigurer$FinalizeTaskAction.execute(TestTaskConfigurer.java:170)
        at org.gradle.testretry.internal.config.TestTaskConfigurer$FinalizeTaskAction.execute(TestTaskConfigurer.java:164)
        at org.gradle.testretry.internal.config.TestTaskConfigurer$ConditionalTaskAction.execute(TestTaskConfigurer.java:159)
        at org.gradle.testretry.internal.config.TestTaskConfigurer$ConditionalTaskAction.execute(TestTaskConfigurer.java:144)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:835)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:808)
        at org.gradle.api.internal.tasks.execution.TaskExecution$3.run(TaskExecution.java:248)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:73)
        at org.gradle.api.internal.tasks.execution.TaskExecution.executeAction(TaskExecution.java:233)
        at org.gradle.api.internal.tasks.execution.TaskExecution.executeActions(TaskExecution.java:216)
        at org.gradle.api.internal.tasks.execution.TaskExecution.executeWithPreviousOutputFiles(TaskExecution.java:199)
        at org.gradle.api.internal.tasks.execution.TaskExecution.execute(TaskExecution.java:166)
        at org.gradle.internal.execution.steps.ExecuteStep.executeInternal(ExecuteStep.java:105)
        at org.gradle.internal.execution.steps.ExecuteStep.access$000(ExecuteStep.java:44)
        at org.gradle.internal.execution.steps.ExecuteStep$1.call(ExecuteStep.java:59)
        at org.gradle.internal.execution.steps.ExecuteStep$1.call(ExecuteStep.java:56)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:199)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:78)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:56)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:44)
        at org.gradle.internal.execution.steps.RemovePreviousOutputsStep.execute(RemovePreviousOutputsStep.java:67)
        at org.gradle.internal.execution.steps.RemovePreviousOutputsStep.execute(RemovePreviousOutputsStep.java:37)
        at org.gradle.internal.execution.steps.CancelExecutionStep.execute(CancelExecutionStep.java:41)
        at org.gradle.internal.execution.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:74)
        at org.gradle.internal.execution.steps.TimeoutStep.execute(TimeoutStep.java:55)
        at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:50)
        at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:28)
        at org.gradle.internal.execution.steps.CaptureStateAfterExecutionStep.executeDelegateBroadcastingChanges(CaptureStateAfterExecutionStep.java:100)
        at org.gradle.internal.execution.steps.CaptureStateAfterExecutionStep.execute(CaptureStateAfterExecutionStep.java:72)
        at org.gradle.internal.execution.steps.CaptureStateAfterExecutionStep.execute(CaptureStateAfterExecutionStep.java:50)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:40)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:29)
        at org.gradle.internal.execution.steps.BuildCacheStep.executeWithoutCache(BuildCacheStep.java:179)
        at org.gradle.internal.execution.steps.BuildCacheStep.lambda$execute$1(BuildCacheStep.java:70)
        at org.gradle.internal.Either$Right.fold(Either.java:175)
        at org.gradle.internal.execution.caching.CachingState.fold(CachingState.java:59)
        at org.gradle.internal.execution.steps.BuildCacheStep.execute(BuildCacheStep.java:68)
        at org.gradle.internal.execution.steps.BuildCacheStep.execute(BuildCacheStep.java:46)
        at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:36)
        at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:25)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:36)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:22)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:91)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$2(SkipUpToDateStep.java:55)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:55)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:37)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:65)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:36)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:37)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:27)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:77)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:38)
        at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:108)
        at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:55)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:71)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:45)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.executeWithNonEmptySources(SkipEmptyWorkStep.java:177)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:86)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:53)
        at org.gradle.internal.execution.steps.RemoveUntrackedExecutionStateStep.execute(RemoveUntrackedExecutionStateStep.java:32)
        at org.gradle.internal.execution.steps.RemoveUntrackedExecutionStateStep.execute(RemoveUntrackedExecutionStateStep.java:21)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsStartedStep.execute(MarkSnapshottingInputsStartedStep.java:38)
        at org.gradle.internal.execution.steps.LoadPreviousExecutionStateStep.execute(LoadPreviousExecutionStateStep.java:36)
        at org.gradle.internal.execution.steps.LoadPreviousExecutionStateStep.execute(LoadPreviousExecutionStateStep.java:23)
        at org.gradle.internal.execution.steps.CleanupStaleOutputsStep.execute(CleanupStaleOutputsStep.java:75)
        at org.gradle.internal.execution.steps.CleanupStaleOutputsStep.execute(CleanupStaleOutputsStep.java:41)
        at org.gradle.internal.execution.steps.ExecuteWorkBuildOperationFiringStep.lambda$execute$2(ExecuteWorkBuildOperationFiringStep.java:66)
        at org.gradle.internal.execution.steps.ExecuteWorkBuildOperationFiringStep.execute(ExecuteWorkBuildOperationFiringStep.java:66)
        at org.gradle.internal.execution.steps.ExecuteWorkBuildOperationFiringStep.execute(ExecuteWorkBuildOperationFiringStep.java:38)
        at org.gradle.internal.execution.steps.AssignWorkspaceStep.lambda$execute$0(AssignWorkspaceStep.java:32)
        at org.gradle.api.internal.tasks.execution.TaskExecution$4.withWorkspace(TaskExecution.java:293)
        at org.gradle.internal.execution.steps.AssignWorkspaceStep.execute(AssignWorkspaceStep.java:30)
        at org.gradle.internal.execution.steps.AssignWorkspaceStep.execute(AssignWorkspaceStep.java:21)
        at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:37)
        at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:27)
        at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:47)
        at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:34)
        at org.gradle.internal.execution.impl.DefaultExecutionEngine$1.execute(DefaultExecutionEngine.java:64)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:145)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:134)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:74)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:199)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:78)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:42)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:331)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:318)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.lambda$execute$0(DefaultTaskExecutionGraph.java:314)
        at org.gradle.internal.operations.CurrentBuildOperationRef.with(CurrentBuildOperationRef.java:80)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:314)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:303)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:463)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:380)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
```

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/JUnitTestEngineListener.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/JUnitTestEngineListener.kt
@@ -119,9 +119,13 @@ class JUnitTestEngineListener(
       logger.log { "specStarted ${ref.kclass}" }
       try {
 
-         // descriptor must definitely exist for a started spec
-         val descriptor = findTestDescriptorForSpec(root, ref.kclass.toDescriptor())
-         descriptors[ref.kclass.toDescriptor()] = descriptor ?: error("Could not find TestDescriptor for ${ref.kclass}")
+         var descriptor = findTestDescriptorForSpec(root, ref.kclass.toDescriptor())
+         if (descriptor == null) {
+            descriptor = createSpecTestDescriptor(root, ref.kclass.toDescriptor(), ref.kclass.bestName())
+            root.addChild(descriptor)
+            listener.dynamicTestRegistered(descriptor)
+         }
+         descriptors[ref.kclass.toDescriptor()] = descriptor
 
          logger.log { Pair(ref.kclass.bestName(), "executionStarted $descriptor") }
          listener.executionStarted(descriptor)

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/JUnitTestEngineListener.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/JUnitTestEngineListener.kt
@@ -19,9 +19,7 @@ import io.kotest.engine.test.names.DisplayNameFormatting
 import org.junit.platform.engine.EngineExecutionListener
 import org.junit.platform.engine.TestDescriptor
 import org.junit.platform.engine.TestExecutionResult
-import org.junit.platform.engine.UniqueId
 import org.junit.platform.engine.support.descriptor.EngineDescriptor
-import org.junit.platform.engine.support.descriptor.MethodSource
 import kotlin.reflect.KClass
 
 /**
@@ -122,7 +120,7 @@ class JUnitTestEngineListener(
       try {
 
          // descriptor must definitely exist for a started spec
-         val descriptor = root.getSpecTestDescriptor(ref.kclass.toDescriptor())
+         val descriptor = findTestDescriptorForSpec(root, ref.kclass.toDescriptor())
          descriptors[ref.kclass.toDescriptor()] = descriptor ?: error("Could not find TestDescriptor for ${ref.kclass}")
 
          logger.log { Pair(ref.kclass.bestName(), "executionStarted $descriptor") }
@@ -143,7 +141,7 @@ class JUnitTestEngineListener(
          // if we had an error in the spec, we will attach a placeholder error-test to the spec
          // and mark that as failed
          t != null -> {
-            val descriptor = root.getSpecTestDescriptor(ref.kclass.toDescriptor())
+            val descriptor = findTestDescriptorForSpec(root, ref.kclass.toDescriptor())
                ?: error("Could not find TestDescriptor for ${ref.kclass}")
             addPlaceholderTest(descriptor, t, ref.kclass)
             logger.log { Pair(ref.kclass.bestName(), "executionFinished: $descriptor $t") }
@@ -151,7 +149,7 @@ class JUnitTestEngineListener(
          }
 
          else -> {
-            val descriptor = root.getSpecTestDescriptor(ref.kclass.toDescriptor())
+            val descriptor = findTestDescriptorForSpec(root, ref.kclass.toDescriptor())
             logger.log { Pair(ref.kclass.bestName(), "executionFinished: $descriptor") }
             listener.executionFinished(descriptor, TestExecutionResult.successful())
          }
@@ -166,7 +164,7 @@ class JUnitTestEngineListener(
       // instead of showing all tests minus the one we are running as ignored, we'll just skip
 
       logger.log { Pair(kclass.bestName(), "Spec is being flagged as ignored") }
-      val testDescriptor = root.getSpecTestDescriptor(kclass.toDescriptor())
+      val testDescriptor = findTestDescriptorForSpec(root, kclass.toDescriptor())
       if (testDescriptor != null)
          listener.executionSkipped(testDescriptor, reason)
    }
@@ -268,7 +266,7 @@ class JUnitTestEngineListener(
       // if it was not started, then it must be a Type.TEST (otherwise its children would have started it)
       startTestIfNotStarted(testCase, TestDescriptor.Type.TEST)
 
-      val descriptor = createTestDescriptorWithMethodSource(testCase, TestDescriptor.Type.TEST)
+      val descriptor = createTestDescriptorWithMethodSource(root, testCase, TestDescriptor.Type.TEST, formatter)
 
       logger.log { Pair(testCase.name.name, "executionFinished: $descriptor") }
       listener.executionFinished(descriptor, result.toTestExecutionResult())
@@ -283,7 +281,7 @@ class JUnitTestEngineListener(
 
       // like all tests, an ignored test should be registered first
       // ignored test should be a TEST type, because an ignored test will never have child tests.
-      val testDescriptor = createTestDescriptorWithMethodSource(testCase, TestDescriptor.Type.TEST)
+      val testDescriptor = createTestDescriptorWithMethodSource(root, testCase, TestDescriptor.Type.TEST, formatter)
       attachToParent(testCase, testDescriptor)
 
       logger.log { Pair(testCase.name.name, "Registering dynamic test: $testDescriptor") }
@@ -312,7 +310,7 @@ class JUnitTestEngineListener(
    private fun startTestIfNotStarted(testCase: TestCase, type: TestDescriptor.Type) {
       if (!startedTests.contains(testCase.descriptor)) {
 
-         val testDescriptor = createTestDescriptorWithMethodSource(testCase, type)
+         val testDescriptor = createTestDescriptorWithMethodSource(root, testCase, type, formatter)
          attachToParent(testCase, testDescriptor)
          descriptors[testCase.descriptor] = testDescriptor
 
@@ -333,28 +331,6 @@ class JUnitTestEngineListener(
       val p = descriptors[parent] ?: error("No parent found: ${parent.id.value}")
       p.addChild(testDescriptor)
    }
-
-   private fun createTestDescriptorWithMethodSource(
-      testCase: TestCase,
-      type: TestDescriptor.Type,
-   ): TestDescriptor {
-      val id = root.deriveTestUniqueId(testCase.descriptor)
-      val testDescriptor = createTestTestDescriptor(
-         id = id,
-         displayName = formatter.format(testCase),
-         type = type,
-         // gradle-junit-platform hides tests if we don't send a source at all
-         // surefire-junit-platform (maven) needs a MethodSource in order to separate test cases from each other
-         // and produce more correct XML report with test case name.
-         source = getMethodSource(testCase.spec::class, id),
-      )
-      return testDescriptor
-   }
-
-   private fun getMethodSource(kclass: KClass<*>, id: UniqueId): MethodSource = MethodSource.from(
-      /* className = */ kclass.qualifiedName,
-      /* methodName = */ id.segments.filter { it.type == Segment.Test.value }.joinToString("/") { it.value }
-   )
 
    /**
     * Registers placeholder specs and marks them as failed for each throwable.

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/JUnitTestEngineListener.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/JUnitTestEngineListener.kt
@@ -70,6 +70,11 @@ import kotlin.reflect.KClass
  * Call addChild _before_ registering test otherwise will appear in the display out of order.
  * Must start tests after their parent or they can go missing.
  * Sibling containers can start and finish in parallel.
+ *
+ * Observations from 1.13.4
+ *
+ * TestDescriptor.Type.CONTAINER that do not contain TESTs are ignored in intellij's tree output
+ *
  */
 class JUnitTestEngineListener(
    private val listener: EngineExecutionListener,
@@ -79,7 +84,7 @@ class JUnitTestEngineListener(
 
    private val logger = Logger(JUnitTestEngineListener::class)
 
-   // contains a mapping of junit TestDescriptor's, so we can find previously registered tests
+   // JUnit TestDescriptors are mutable and contain children added dynamically, so we need to keep track of them
    private val descriptors = mutableMapOf<Descriptor, TestDescriptor>()
 
    private val startedTests = mutableSetOf<Descriptor.TestDescriptor>()
@@ -120,11 +125,13 @@ class JUnitTestEngineListener(
       try {
 
          var descriptor = findTestDescriptorForSpec(root, ref.kclass.toDescriptor())
+
          if (descriptor == null) {
             descriptor = createSpecTestDescriptor(root, ref.kclass.toDescriptor(), ref.kclass.bestName())
             root.addChild(descriptor)
             listener.dynamicTestRegistered(descriptor)
          }
+
          descriptors[ref.kclass.toDescriptor()] = descriptor
 
          logger.log { Pair(ref.kclass.bestName(), "executionStarted $descriptor") }
@@ -146,7 +153,7 @@ class JUnitTestEngineListener(
          // and mark that as failed
          t != null -> {
             val descriptor = findTestDescriptorForSpec(root, ref.kclass.toDescriptor())
-               ?: error("Could not find TestDescriptor for ${ref.kclass}")
+               ?: error("Could not find TestDescriptor for spec ${ref.kclass}")
             addPlaceholderTest(descriptor, t, ref.kclass)
             logger.log { Pair(ref.kclass.bestName(), "executionFinished: $descriptor $t") }
             listener.executionFinished(descriptor, TestExecutionResult.failed(t))
@@ -154,6 +161,7 @@ class JUnitTestEngineListener(
 
          else -> {
             val descriptor = findTestDescriptorForSpec(root, ref.kclass.toDescriptor())
+               ?: error("Could not find TestDescriptor for spec ${ref.kclass}")
             logger.log { Pair(ref.kclass.bestName(), "executionFinished: $descriptor") }
             listener.executionFinished(descriptor, TestExecutionResult.successful())
          }
@@ -172,24 +180,6 @@ class JUnitTestEngineListener(
       if (testDescriptor != null)
          listener.executionSkipped(testDescriptor, reason)
    }
-
-//   private fun markSpecStarted(kclass: KClass<*>): TestDescriptor {
-//      return try {
-//
-//         log { "Getting TestDescriptor for $kclass" }
-//         val descriptor = getSpecDescriptor(kclass)
-//
-//         logger.log { Pair(kclass.bestName(), "Spec executionStarted $descriptor") }
-//         listener.executionStarted(descriptor)
-//
-//         started = true
-//         descriptor
-//
-//      } catch (t: Throwable) {
-//         logger.log { Pair(kclass.bestName(), "Error marking spec as started $t") }
-//         throw t
-//      }
-//   }
 
    private fun reset() {
       results.clear()
@@ -222,25 +212,12 @@ class JUnitTestEngineListener(
 
    override suspend fun testStarted(testCase: TestCase) {
 
-      // We want to wait to notify junit, this is because gradle doesn't work properly with the junit test types.
-      // Ideally, we'd just set everything to CONTAINER_AND_TEST, which is supposed to mean a test can contain
-      // other tests as well as being a test itself, which is exactly how Kotest views tests, but unfortunately
-      // it's not supported by gradle :(
-      //
-      // So we need to not start tests until we can determine if they are a TEST or a CONTAINER.
-      // Once a nested test starts, we can start the parent, since we know the parent is definitely a CONTAINER
-      // at that point.
-      //
-      // Leaf tests must wait until they complete so we know there were no child tests.
-      //
-      // Further annoyance is that junit doesn't give us a way to specify test duration
-      // (instead it just calculates it itself from the time between marking a test as started and marking
-      // it as finished), so we end up having all leaf tests as 0ms
-      //
-      // Therefore, our workaround is to just add the execution time into the test name.
+      // JUnit has TEST and CONTAINER types, but intellij will not render a CONTAINER unless it has a child.
+      // Since Kotest allows you to have containers that do not contain tests, we must wait to see if a container
+      // has any children before we register it as a container. If it does not, then we will register it as a test
+      // once it completes. Therefore the testStarted listener must wait to register a test.
 
       logger.log { Pair(testCase.name.name, "test started") }
-
 
       // if this test has a parent, we can mark that parent as started, because it's definitely not a leaf
       if (testCase.parent != null)
@@ -308,7 +285,7 @@ class JUnitTestEngineListener(
    }
 
    /**
-    * If the given testCase has not yet been marked as tested, then now we register it and start it.
+    * If the given testCase has not yet been marked as started, then now we register it and start it.
     * This method is only invoked for parents, so we know the type is always CONTAINER.
     */
    private fun startTestIfNotStarted(testCase: TestCase, type: TestDescriptor.Type) {

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/KotestEngineDescriptor.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/KotestEngineDescriptor.kt
@@ -14,9 +14,9 @@ class KotestEngineDescriptor internal constructor(
    val classes: List<KClass<out Spec>>,
    val extensions: List<Extension>,
 ) : EngineDescriptor(id, KotestJunitPlatformTestEngine.ENGINE_NAME) {
-   // Only reports dynamic children (see ExtensionExceptionExtractor) if there are no test classes to run,
-   // so that we can add a placeholder test for the error
-   override fun mayRegisterTests(): Boolean = classes.isEmpty()
+   // we add tests dynamically to the engine when there's an engine exception to be reported, and also
+   // some plugins like the gradle test-retry-plugin prune the root test descriptor so they are re-registered
+   override fun mayRegisterTests(): Boolean = true
 }
 
 internal data class EngineDescriptorBuilder(

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/descriptors.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/descriptors.kt
@@ -70,7 +70,7 @@ internal fun createTestDescriptorWithMethodSource(
    type: TestDescriptor.Type,
    formatter: DisplayNameFormatting,
 ): TestDescriptor {
-   val id = createTestUniqueId(root.uniqueId, testCase.descriptor)
+   val id = createUniqueIdForTest(root.uniqueId, testCase.descriptor)
    val testDescriptor = createTestTestDescriptor(
       id = id,
       displayName = formatter.format(testCase),

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/descriptors.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/descriptors.kt
@@ -14,8 +14,10 @@ import kotlin.jvm.optionals.getOrNull
 import kotlin.reflect.KClass
 
 /**
- * Returns the [org.junit.platform.engine.TestDescriptor] corresponding to the given spec.
- * Specs are always registered when the test suite is created, so this is expected to never fail.
+ * Finds and returns the [org.junit.platform.engine.TestDescriptor] corresponding to the
+ * given [Descriptor.SpecDescriptor] that was previously added to the engine.
+ *
+ * If the engine descriptor does not contain the spec descriptor, then null is returned.
  */
 internal fun findTestDescriptorForSpec(root: EngineDescriptor, descriptor: Descriptor.SpecDescriptor): TestDescriptor? {
    val id = createUniqueIdForSpec(root, descriptor.id)

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/descriptors.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/descriptors.kt
@@ -18,7 +18,7 @@ import kotlin.reflect.KClass
  * Specs are always registered when the test suite is created, so this is expected to never fail.
  */
 internal fun findTestDescriptorForSpec(root: EngineDescriptor, descriptor: Descriptor.SpecDescriptor): TestDescriptor? {
-   val id = root.deriveSpecUniqueId(descriptor.id)
+   val id = createUniqueIdForSpec(root, descriptor.id)
    return root.findByUniqueId(id).getOrNull()
 }
 
@@ -31,7 +31,7 @@ internal fun createSpecTestDescriptor(
    descriptor: Descriptor.SpecDescriptor,
    displayName: String,
 ): TestDescriptor {
-   val id = root.deriveSpecUniqueId(descriptor.id)
+   val id = createUniqueIdForSpec(root, descriptor.id)
    val source = ClassSource.from(descriptor.id.value)
    return object : AbstractTestDescriptor(id, displayName, source) {
       override fun getType(): TestDescriptor.Type = TestDescriptor.Type.CONTAINER

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/descriptors.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/descriptors.kt
@@ -1,21 +1,25 @@
 package io.kotest.runner.junit.platform
 
 import io.kotest.core.descriptors.Descriptor
+import io.kotest.core.test.TestCase
+import io.kotest.engine.test.names.DisplayNameFormatting
 import org.junit.platform.engine.TestDescriptor
 import org.junit.platform.engine.TestSource
 import org.junit.platform.engine.UniqueId
 import org.junit.platform.engine.support.descriptor.AbstractTestDescriptor
 import org.junit.platform.engine.support.descriptor.ClassSource
 import org.junit.platform.engine.support.descriptor.EngineDescriptor
+import org.junit.platform.engine.support.descriptor.MethodSource
 import kotlin.jvm.optionals.getOrNull
+import kotlin.reflect.KClass
 
 /**
  * Returns the [org.junit.platform.engine.TestDescriptor] corresponding to the given spec.
  * Specs are always registered when the test suite is created, so this is expected to never fail.
  */
-internal fun EngineDescriptor.getSpecTestDescriptor(descriptor: Descriptor.SpecDescriptor): TestDescriptor? {
-   val id = deriveSpecUniqueId(descriptor.id)
-   return findByUniqueId(id).getOrNull()
+internal fun findTestDescriptorForSpec(root: EngineDescriptor, descriptor: Descriptor.SpecDescriptor): TestDescriptor? {
+   val id = root.deriveSpecUniqueId(descriptor.id)
+   return root.findByUniqueId(id).getOrNull()
 }
 
 /**
@@ -23,11 +27,11 @@ internal fun EngineDescriptor.getSpecTestDescriptor(descriptor: Descriptor.SpecD
  * This descriptor needs to be added to the parent engine descriptor.
  */
 internal fun createSpecTestDescriptor(
-   engine: EngineDescriptor,
+   root: EngineDescriptor,
    descriptor: Descriptor.SpecDescriptor,
    displayName: String,
 ): TestDescriptor {
-   val id = engine.deriveSpecUniqueId(descriptor.id)
+   val id = root.deriveSpecUniqueId(descriptor.id)
    val source = ClassSource.from(descriptor.id.value)
    return object : AbstractTestDescriptor(id, displayName, source) {
       override fun getType(): TestDescriptor.Type = TestDescriptor.Type.CONTAINER
@@ -57,3 +61,27 @@ internal fun createTestTestDescriptor(
    override fun getType(): TestDescriptor.Type = type
    override fun mayRegisterTests(): Boolean = type == TestDescriptor.Type.CONTAINER
 }
+
+internal fun createTestDescriptorWithMethodSource(
+   root: EngineDescriptor,
+   testCase: TestCase,
+   type: TestDescriptor.Type,
+   formatter: DisplayNameFormatting,
+): TestDescriptor {
+   val id = root.deriveTestUniqueId(testCase.descriptor)
+   val testDescriptor = createTestTestDescriptor(
+      id = id,
+      displayName = formatter.format(testCase),
+      type = type,
+      // gradle-junit-platform hides tests if we don't send a source at all
+      // surefire-junit-platform (maven) needs a MethodSource in order to separate test cases from each other
+      // and produce more correct XML report with test case name.
+      source = getMethodSource(testCase.spec::class, id),
+   )
+   return testDescriptor
+}
+
+internal fun getMethodSource(kclass: KClass<*>, id: UniqueId): MethodSource = MethodSource.from(
+   /* className = */ kclass.qualifiedName,
+   /* methodName = */ id.segments.filter { it.type == Segment.Test.value }.joinToString("/") { it.value }
+)

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/descriptors.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/descriptors.kt
@@ -20,7 +20,7 @@ import kotlin.reflect.KClass
  * If the engine descriptor does not contain the spec descriptor, then null is returned.
  */
 internal fun findTestDescriptorForSpec(root: EngineDescriptor, descriptor: Descriptor.SpecDescriptor): TestDescriptor? {
-   val id = createUniqueIdForSpec(root, descriptor.id)
+   val id = createUniqueIdForSpec(root.uniqueId, descriptor.id)
    return root.findByUniqueId(id).getOrNull()
 }
 
@@ -33,7 +33,7 @@ internal fun createSpecTestDescriptor(
    descriptor: Descriptor.SpecDescriptor,
    displayName: String,
 ): TestDescriptor {
-   val id = createUniqueIdForSpec(root, descriptor.id)
+   val id = createUniqueIdForSpec(root.uniqueId, descriptor.id)
    val source = ClassSource.from(descriptor.id.value)
    return object : AbstractTestDescriptor(id, displayName, source) {
       override fun getType(): TestDescriptor.Type = TestDescriptor.Type.CONTAINER
@@ -70,7 +70,7 @@ internal fun createTestDescriptorWithMethodSource(
    type: TestDescriptor.Type,
    formatter: DisplayNameFormatting,
 ): TestDescriptor {
-   val id = root.deriveTestUniqueId(testCase.descriptor)
+   val id = createTestUniqueId(root.uniqueId, testCase.descriptor)
    val testDescriptor = createTestTestDescriptor(
       id = id,
       displayName = formatter.format(testCase),

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/uniqueids.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/uniqueids.kt
@@ -29,18 +29,23 @@ internal sealed class Segment {
 }
 
 /**
- * The created [UniqueId] will have segment type [Segment.Spec] and will use the descriptor id.
+ * Creates a [UniqueId] for a spec from the given [EngineDescriptor] and [DescriptorId].
+ *
+ * The created id will have segment type [Segment.Spec].
  */
-internal fun createUniqueIdForSpec(root: EngineDescriptor, id: DescriptorId): UniqueId =
-   root.uniqueId.append(Segment.Spec.value, id.value)
+internal fun createUniqueIdForSpec(engineId: UniqueId, id: DescriptorId): UniqueId =
+   engineId.append(Segment.Spec.value, id.value)
 
 /**
- * The created [UniqueId] will have segment type [Segment.Test] and will use the descriptor id.
+ * Creates a [UniqueId] for a test from the given [EngineDescriptor] and [Descriptor.TestDescriptor].
+ *
+ * The created id will have segment type [Segment.Test], and any parent tests plus the spec will
+ * be prepended to the created id.
  */
-internal fun EngineDescriptor.deriveTestUniqueId(descriptor: Descriptor): UniqueId {
-   return when (descriptor) {
-      is Descriptor.SpecDescriptor -> createUniqueIdForSpec(this, descriptor.id)
-      is Descriptor.TestDescriptor -> deriveTestUniqueId(descriptor.parent)
-         .append(Segment.Test.value, descriptor.id.value)
+internal fun createTestUniqueId(engineId: UniqueId, descriptor: Descriptor.TestDescriptor): UniqueId {
+   val parentDescriptor = when (val parent = descriptor.parent) {
+      is Descriptor.SpecDescriptor -> createUniqueIdForSpec(engineId, parent.id)
+      is Descriptor.TestDescriptor -> createTestUniqueId(engineId, parent)
    }
+   return parentDescriptor.append(Segment.Test.value, descriptor.id.value)
 }

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/uniqueids.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/uniqueids.kt
@@ -31,10 +31,10 @@ internal fun createUniqueIdForSpec(engineId: UniqueId, id: DescriptorId): Unique
  * The created id will have segment type [Segment.Test], and any parent tests plus the spec will
  * be prepended to the created id.
  */
-internal fun createTestUniqueId(engineId: UniqueId, descriptor: Descriptor.TestDescriptor): UniqueId {
+internal fun createUniqueIdForTest(engineId: UniqueId, descriptor: Descriptor.TestDescriptor): UniqueId {
    val parentDescriptor = when (val parent = descriptor.parent) {
       is Descriptor.SpecDescriptor -> createUniqueIdForSpec(engineId, parent.id)
-      is Descriptor.TestDescriptor -> createTestUniqueId(engineId, parent)
+      is Descriptor.TestDescriptor -> createUniqueIdForTest(engineId, parent)
    }
    return parentDescriptor.append(Segment.Test.value, descriptor.id.value)
 }

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/uniqueids.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/uniqueids.kt
@@ -31,15 +31,15 @@ internal sealed class Segment {
 /**
  * The created [UniqueId] will have segment type [Segment.Spec] and will use the descriptor id.
  */
-internal fun EngineDescriptor.deriveSpecUniqueId(id: DescriptorId): UniqueId =
-   uniqueId.append(Segment.Spec.value, id.value)
+internal fun createUniqueIdForSpec(root: EngineDescriptor, id: DescriptorId): UniqueId =
+   root.uniqueId.append(Segment.Spec.value, id.value)
 
 /**
  * The created [UniqueId] will have segment type [Segment.Test] and will use the descriptor id.
  */
 internal fun EngineDescriptor.deriveTestUniqueId(descriptor: Descriptor): UniqueId {
    return when (descriptor) {
-      is Descriptor.SpecDescriptor -> deriveSpecUniqueId(descriptor.id)
+      is Descriptor.SpecDescriptor -> createUniqueIdForSpec(this, descriptor.id)
       is Descriptor.TestDescriptor -> deriveTestUniqueId(descriptor.parent)
          .append(Segment.Test.value, descriptor.id.value)
    }

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/uniqueids.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmMain/kotlin/io/kotest/runner/junit/platform/uniqueids.kt
@@ -5,17 +5,6 @@ import io.kotest.core.descriptors.DescriptorId
 import org.junit.platform.engine.UniqueId
 import org.junit.platform.engine.support.descriptor.EngineDescriptor
 
-/**
- * Returns a new [UniqueId] by appending this [descriptor] to the receiver.
- */
-internal fun UniqueId.append(descriptor: Descriptor): UniqueId {
-   val segment = when (descriptor) {
-      is Descriptor.SpecDescriptor -> Segment.Spec
-      is Descriptor.TestDescriptor -> Segment.Test
-   }
-   return this.append(segment.value, descriptor.id.value)
-}
-
 internal sealed class Segment {
    abstract val value: String
 

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit/platform/UniqueIdTest.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit/platform/UniqueIdTest.kt
@@ -4,7 +4,7 @@ import io.kotest.core.descriptors.toDescriptor
 import io.kotest.core.spec.style.FunSpec
 import io.kotest.matchers.shouldBe
 import io.kotest.runner.junit.platform.Segment
-import io.kotest.runner.junit.platform.createTestUniqueId
+import io.kotest.runner.junit.platform.createUniqueIdForTest
 import io.kotest.runner.junit.platform.createUniqueIdForSpec
 import org.junit.platform.engine.UniqueId
 
@@ -20,7 +20,7 @@ class UniqueIdTest : FunSpec() {
       test("createTestUniqueId should handle rooot tests") {
          val root = UniqueId.forEngine("kotest")
          val testDescriptor = UniqueIdTest::class.toDescriptor().append("a")
-         createTestUniqueId(root, testDescriptor) shouldBe
+         createUniqueIdForTest(root, testDescriptor) shouldBe
             root.append(Segment.Spec.value, "com.sksamuel.kotest.runner.junit.platform.UniqueIdTest")
                .append(Segment.Test.value, "a")
       }
@@ -28,7 +28,7 @@ class UniqueIdTest : FunSpec() {
       test("createTestUniqueId should handle nested tests") {
          val root = UniqueId.forEngine("kotest")
          val testDescriptor = UniqueIdTest::class.toDescriptor().append("a").append("b")
-         createTestUniqueId(root, testDescriptor) shouldBe
+         createUniqueIdForTest(root, testDescriptor) shouldBe
             root.append(Segment.Spec.value, "com.sksamuel.kotest.runner.junit.platform.UniqueIdTest")
                .append(Segment.Test.value, "a")
                .append(Segment.Test.value, "b")

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit/platform/UniqueIdTest.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit/platform/UniqueIdTest.kt
@@ -1,0 +1,37 @@
+package com.sksamuel.kotest.runner.junit.platform
+
+import io.kotest.core.descriptors.toDescriptor
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.matchers.shouldBe
+import io.kotest.runner.junit.platform.Segment
+import io.kotest.runner.junit.platform.createTestUniqueId
+import io.kotest.runner.junit.platform.createUniqueIdForSpec
+import org.junit.platform.engine.UniqueId
+
+class UniqueIdTest : FunSpec() {
+   init {
+
+      test("createUniqueIdForSpec") {
+         val root = UniqueId.forEngine("kotest")
+         createUniqueIdForSpec(root, UniqueIdTest::class.toDescriptor().id) shouldBe
+            root.append(Segment.Spec.value, "com.sksamuel.kotest.runner.junit.platform.UniqueIdTest")
+      }
+
+      test("createTestUniqueId should handle rooot tests") {
+         val root = UniqueId.forEngine("kotest")
+         val testDescriptor = UniqueIdTest::class.toDescriptor().append("a")
+         createTestUniqueId(root, testDescriptor) shouldBe
+            root.append(Segment.Spec.value, "com.sksamuel.kotest.runner.junit.platform.UniqueIdTest")
+               .append(Segment.Test.value, "a")
+      }
+
+      test("createTestUniqueId should handle nested tests") {
+         val root = UniqueId.forEngine("kotest")
+         val testDescriptor = UniqueIdTest::class.toDescriptor().append("a").append("b")
+         createTestUniqueId(root, testDescriptor) shouldBe
+            root.append(Segment.Spec.value, "com.sksamuel.kotest.runner.junit.platform.UniqueIdTest")
+               .append(Segment.Test.value, "a")
+               .append(Segment.Test.value, "b")
+      }
+   }
+}

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit/platform/UniqueIdTest.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit/platform/UniqueIdTest.kt
@@ -17,7 +17,7 @@ class UniqueIdTest : FunSpec() {
             root.append(Segment.Spec.value, "com.sksamuel.kotest.runner.junit.platform.UniqueIdTest")
       }
 
-      test("createTestUniqueId should handle rooot tests") {
+      test("createTestUniqueId should handle root tests") {
          val root = UniqueId.forEngine("kotest")
          val testDescriptor = UniqueIdTest::class.toDescriptor().append("a")
          createUniqueIdForTest(root, testDescriptor) shouldBe

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit5/DynamicSpecRegistrationTest.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit5/DynamicSpecRegistrationTest.kt
@@ -1,12 +1,15 @@
 package com.sksamuel.kotest.runner.junit5
 
 import io.kotest.core.annotation.Issue
+import io.kotest.core.descriptors.DescriptorId
 import io.kotest.core.spec.SpecRef
 import io.kotest.core.spec.style.FunSpec
 import io.kotest.engine.test.names.DisplayNameFormatting
+import io.kotest.matchers.nulls.shouldNotBeNull
 import io.kotest.matchers.shouldBe
 import io.kotest.runner.junit.platform.EngineDescriptorBuilder
 import io.kotest.runner.junit.platform.JUnitTestEngineListener
+import io.kotest.runner.junit.platform.createUniqueIdForSpec
 import org.junit.platform.engine.EngineExecutionListener
 import org.junit.platform.engine.TestDescriptor
 import org.junit.platform.engine.TestExecutionResult
@@ -18,23 +21,30 @@ class DynamicSpecRegistrationTest : FunSpec() {
    init {
       test("JUnitTestEngineListener should dynamically register spec if not currently registered") {
 
-         var registered = false
+         var registered: TestDescriptor? = null
+         var started: TestDescriptor? = null
 
          val engineExecutionListener = object : EngineExecutionListener {
             override fun executionFinished(testDescriptor: TestDescriptor, testExecutionResult: TestExecutionResult) {}
             override fun reportingEntryPublished(testDescriptor: TestDescriptor?, entry: ReportEntry?) {}
             override fun executionSkipped(testDescriptor: TestDescriptor?, reason: String?) {}
-            override fun executionStarted(testDescriptor: TestDescriptor?) {}
+            override fun executionStarted(testDescriptor: TestDescriptor?) {
+               started = testDescriptor
+            }
+
             override fun dynamicTestRegistered(testDescriptor: TestDescriptor?) {
-               registered = true
+               registered = testDescriptor
             }
          }
 
-         val root = EngineDescriptorBuilder.builder(UniqueId.forEngine("kotest")).build()
+         val rootId = UniqueId.forEngine("kotest")
+         val root = EngineDescriptorBuilder.builder(rootId).build()
          val listener = JUnitTestEngineListener(engineExecutionListener, root, DisplayNameFormatting(null))
          listener.specStarted(SpecRef.Reference(DynamicSpecRegistrationTest::class))
 
-         registered shouldBe true
+         val specId = createUniqueIdForSpec(root, DescriptorId(DynamicSpecRegistrationTest::class.java.name))
+         registered.shouldNotBeNull().uniqueId shouldBe specId
+         started.shouldNotBeNull().uniqueId shouldBe specId
       }
    }
 }

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit5/DynamicSpecRegistrationTest.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit5/DynamicSpecRegistrationTest.kt
@@ -1,0 +1,40 @@
+package com.sksamuel.kotest.runner.junit5
+
+import io.kotest.core.annotation.Issue
+import io.kotest.core.spec.SpecRef
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.engine.test.names.DisplayNameFormatting
+import io.kotest.matchers.shouldBe
+import io.kotest.runner.junit.platform.EngineDescriptorBuilder
+import io.kotest.runner.junit.platform.JUnitTestEngineListener
+import org.junit.platform.engine.EngineExecutionListener
+import org.junit.platform.engine.TestDescriptor
+import org.junit.platform.engine.TestExecutionResult
+import org.junit.platform.engine.UniqueId
+import org.junit.platform.engine.reporting.ReportEntry
+
+@Issue("https://github.com/kotest/kotest/issues/3770")
+class DynamicSpecRegistrationTest : FunSpec() {
+   init {
+      test("JUnitTestEngineListener should dynamically register spec if not currently registered") {
+
+         var registered = false
+
+         val engineExecutionListener = object : EngineExecutionListener {
+            override fun executionFinished(testDescriptor: TestDescriptor, testExecutionResult: TestExecutionResult) {}
+            override fun reportingEntryPublished(testDescriptor: TestDescriptor?, entry: ReportEntry?) {}
+            override fun executionSkipped(testDescriptor: TestDescriptor?, reason: String?) {}
+            override fun executionStarted(testDescriptor: TestDescriptor?) {}
+            override fun dynamicTestRegistered(testDescriptor: TestDescriptor?) {
+               registered = true
+            }
+         }
+
+         val root = EngineDescriptorBuilder.builder(UniqueId.forEngine("kotest")).build()
+         val listener = JUnitTestEngineListener(engineExecutionListener, root, DisplayNameFormatting(null))
+         listener.specStarted(SpecRef.Reference(DynamicSpecRegistrationTest::class))
+
+         registered shouldBe true
+      }
+   }
+}

--- a/kotest-runner/kotest-runner-junit-platform/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit5/DynamicSpecRegistrationTest.kt
+++ b/kotest-runner/kotest-runner-junit-platform/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit5/DynamicSpecRegistrationTest.kt
@@ -42,7 +42,7 @@ class DynamicSpecRegistrationTest : FunSpec() {
          val listener = JUnitTestEngineListener(engineExecutionListener, root, DisplayNameFormatting(null))
          listener.specStarted(SpecRef.Reference(DynamicSpecRegistrationTest::class))
 
-         val specId = createUniqueIdForSpec(root, DescriptorId(DynamicSpecRegistrationTest::class.java.name))
+         val specId = createUniqueIdForSpec(root.uniqueId, DescriptorId(DynamicSpecRegistrationTest::class.java.name))
          registered.shouldNotBeNull().uniqueId shouldBe specId
          started.shouldNotBeNull().uniqueId shouldBe specId
       }

